incubator/virtualcluster/pkg/syncer/syncer.go

```diff
@@ -74,37 +81,82 @@ func New(
 	go func() {
-		s.controllerManager.Start(signals.SetupSignalHandler())
+		s.controllerManager.Start(s.stopChan)
 	}()
 }
 
-func (s *Syncer) onVirtualClusterAdd(obj interface{}) {
-	vc, ok := obj.(*v1alpha1.Virtualcluster)
-	if !ok {
-		klog.Errorf("cannot convert to *v1alpha1.VirtualCluster: %v", obj)
-		return
+func isTLSTimeout(err error) bool {
+	if err, ok := err.(net.Error); ok && err.Timeout() {
+		return true
 	}
+	return false
+}
 
-	klog.Infof("handle virtual cluster %s add event", vc.Name)
-
-	// Build cluster admin client based on admin.kubeconfig secret
-	adminKubeconfigSecret, err := s.secretClient.Secrets(vc.Namespace).Get(KubeconfigAdmin, metav1.GetOptions{})
-	if err != nil {
-		klog.Errorf("failed to get admin.kubeconfig secret for virtual cluster %s: %v", vc.Name, err)
+// onVirtualClusterUpdate checks if a Virtualcluster is starting to run. For a
+// newly started Virtualcluster, we will register it with all resources
+// controllers (e.g. pod, node, etc.).
+func (s *Syncer) onVirtualClusterUpdate(old, new interface{}) {
+	oldVc, ok := old.(*v1alpha1.Virtualcluster)
+	if !ok {
+		klog.Errorf("cannot convert to *v1alpha1.VirtualCluster: %v", old)
 		return
 	}
-	clusterRestConfig, err := clientcmd.RESTConfigFromKubeConfig(adminKubeconfigSecret.Data[KubeconfigAdmin])
-	if err != nil {
-		klog.Errorf("failed to build rest config for virtual cluster %s: %v", vc.Name, err)
+	vc, ok := new.(*v1alpha1.Virtualcluster)
+	if !ok {
+		klog.Errorf("cannot convert to *v1alpha1.VirtualCluster: %v", new)
 		return
 	}
 
-	innerCluster := &cluster.Cluster{
-		Name:   vc.Name,
-		Config: clusterRestConfig,
-	}
-	for _, clusterChangeListener := range listener.Listeners {
-		clusterChangeListener.AddCluster(innerCluster)
+	switch {
+	case vc.Status.Phase == v1alpha1.ClusterPending:
+		klog.Infof("Virtualcluster(%s) is pending", vc.Name)
+	case vc.Status.Phase == v1alpha1.ClusterRunning &&
+		oldVc.Status.Phase != v1alpha1.ClusterRunning:
+		// if a Virtualcluster is starting to run, build a cluster admin client
+		// based on the admin-kubeconfig secret of the Virtualcluster.
+		klog.Infof("Virtualcluster(%s) is running", vc.Name)
+		adminKubeconfigSecret, err :=
+			s.secretClient.Secrets(vc.Namespace).Get(KubeconfigAdmin, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("failed to get secret (%s) for virtual cluster %s: %v",
+				KubeconfigAdmin, vc.Name, err)
+			return
+		}
+		clusterRestConfig, err :=
+			clientcmd.RESTConfigFromKubeConfig(adminKubeconfigSecret.Data[KubeconfigAdmin])
+		if err != nil {
+			klog.Errorf("failed to build rest config for virtual cluster %s: %v", vc.Name, err)
+			return
+		}
+		innerCluster := &cluster.Cluster{
+			Name:   vc.Name,
+			Config: clusterRestConfig,
+		}
+
+		// for each resource type of the newly added Virtualcluster, we add a listener
+		for _, clusterChangeListener := range listener.Listeners {
+		RETRY:
+			if err := clusterChangeListener.AddCluster(innerCluster); err != nil {
```

Comment:
I think the `multiclustercontroller` use map to store clusters (i.e. map[cluster]struct{}). Is using map to prevent duplicates? Please correct me if I am wrong.